File: C#/Allergies/Allergies.cs
using System;
using System.Collections.Generic;

public enum Allergen
{
    Eggs = 1,
    Peanuts = 2,
    Shellfish = 4,
    Strawberries = 8,
    Tomatoes = 16,
    Chocolate = 32,
    Pollen = 64,
    Cats = 128,
}

public class Allergies
{
    List<Allergen> _allergies = new List<Allergen>();

    public Allergies(int mask)
    {
        _allergies = Enum.GetValues(typeof(Allergen))
                         .Cast<Allergen>()
                         .Where(x => ((int)x & mask) != 0)
                         .ToList();

        //foreach (var allegen in Enum.GetValues(typeof(Allergen)))
        //{
        //    var result = (mask & (int)allegen);

        //    if (result != 0)
        //    {
        //        _allergies.Add((Allergen)allegen);
        //    }
        //}
    }

    public bool IsAllergicTo(Allergen allergen) => _allergies.Contains(allergen);

    public Allergen[] List() => _allergies.ToArray();
}

File: TypeScript/allergies.ts
interface Allergy {
    [key: string]: string;
}
const ALLERGY_LIST: Allergy = {
    '1': 'eggs',
    '2': 'peanuts',
    '4': 'shellfish',
    '8': 'strawberries',
    '16': 'tomatoes', 
    '32': 'chocolate', 
    '64': 'pollen', 
    '128': 'cats',
}

export class Allergies {
    private _score: number;
    private _list: string[];

    constructor(allergenIndex: number) {
        this._score = allergenIndex;

        // Construct the allergy list
        this._list = [];
        for(let key in ALLERGY_LIST) {
            if(this._score & Number(key)) {
                this._list.push(ALLERGY_LIST[key]);
            }
        }
    }
  
    list(): string[] {
        return this._list;
    }
  
    allergicTo(allergen: string) {
        return (this._list.includes(allergen));
    }
}

let n = new Allergies(9);
console.log(n);
console.log(n.list());

console.log(n.allergicTo('eggs'));
console.log(n.allergicTo('shellfish'));